File: Games/Uzta/src/main/java/jokatu/game/games/uzta/game/MainStageSelectEdgeInputAcceptor.java
package jokatu.game.games.uzta.game;

import jokatu.game.event.PublicGameEvent;
import jokatu.game.games.uzta.graph.LineSegment;
import jokatu.game.games.uzta.graph.NodeType;
import jokatu.game.games.uzta.graph.UztaGraph;
import jokatu.game.games.uzta.input.SelectEdgeInput;
import jokatu.game.games.uzta.player.UztaPlayer;
import jokatu.game.input.UnacceptableInputException;
import jokatu.game.turn.TurnManager;
import ophelia.collections.bag.BaseIntegerBag;
import ophelia.collections.bag.HashBag;
import ophelia.collections.bag.ModifiableIntegerBag;
import ophelia.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.Arrays;
import java.util.Random;
import java.util.stream.Collectors;

import static jokatu.game.games.uzta.game.Uzta.DICE_SIZE;

public class MainStageSelectEdgeInputAcceptor extends AbstractSelectEdgeInputAcceptor {

	private final ResourceDistributor resourceDistributor;
	private final Random die;

	MainStageSelectEdgeInputAcceptor(@NotNull UztaGraph graph, @NotNull TurnManager<UztaPlayer> turnManager, @NotNull ResourceDistributor resourceDistributor) {
		super(graph, turnManager);
		this.resourceDistributor = resourceDistributor;

		turnManager.observe(e -> roll());

		die = new Random();
	}

	private void roll() {
		int roll = 1 + die.nextInt(DICE_SIZE);
		UztaPlayer currentPlayer = turnManager.getCurrentPlayer();
		assert currentPlayer != null;
		fireEvent((PublicGameEvent) () -> MessageFormat.format("{0} rolled {1}", currentPlayer.getName(), roll));

		resourceDistributor.distributeResourcesForRoll(roll);
	}

	@Override
	protected void acceptCastInputAndPlayer(@NotNull SelectEdgeInput input, @NotNull UztaPlayer inputter) throws Exception {
		LineSegment edge = getUnownedLineSegment(input, inputter);

		BaseIntegerBag<NodeType> edgeCost = getCost(edge);
		BaseIntegerBag<NodeType> resourcesLeft = inputter.getResourcesLeftAfter(edgeCost);
		if (resourcesLeft.isLacking()) {
			throw new UnacceptableInputException(
					"You can''t afford that edge.  You still need {0}.",
					resourcesLeft.stream()
							.filter(pair -> pair.getRight() < 0)
							.map(this::presentNeededResources)
							.collect(Collectors.joining(", "))
			);
		}

		setOwner(edge, inputter);
	}

	@NotNull
	private String presentNeededResources(Pair<NodeType, Integer> pair) {
		Integer numberNeeded = -pair.getRight();
		return pair.getLeft().getNumber(numberNeeded);
	}

	private BaseIntegerBag<NodeType> getCost(@NotNull LineSegment edge) {
		ModifiableIntegerBag<NodeType> bag = new HashBag<>();
		edge.forEach(node -> bag.modifyNumberOf(node.getType(), node.getValues().size()));
		Arrays.stream(NodeType.values()).forEach(bag::addOne);
		return bag;
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/game/MainStageSelectEdgeInputAcceptor.kt
package jokatu.game.games.uzta.game

import jokatu.game.event.StandardPublicGameEvent
import jokatu.game.games.uzta.graph.LineSegment
import jokatu.game.games.uzta.graph.NodeType
import jokatu.game.games.uzta.graph.UztaGraph
import jokatu.game.games.uzta.input.SelectEdgeInput
import jokatu.game.games.uzta.player.UztaPlayer
import jokatu.game.input.UnacceptableInputException
import jokatu.game.turn.TurnManager
import ophelia.collections.bag.BaseIntegerBag
import ophelia.collections.bag.HashBag
import ophelia.tuple.Pair
import java.util.*
import java.util.stream.Collectors


class MainStageSelectEdgeInputAcceptor internal constructor(
		graph: UztaGraph,
		turnManager: TurnManager<UztaPlayer>,
		private val resourceDistributor: ResourceDistributor
) : AbstractSelectEdgeInputAcceptor(graph, turnManager) {

	private val die: Random

	init {
		turnManager.observe { _ -> roll() }

		die = Random()
	}

	private fun roll() {
		val roll = 1 + die.nextInt(Uzta.DICE_SIZE)
		val currentPlayer = turnManager.currentPlayer!!
		fireEvent(StandardPublicGameEvent("${currentPlayer.name} rolled $roll"))

		resourceDistributor.distributeResourcesForRoll(roll)
	}

	@Throws(Exception::class)
	override fun acceptCastInputAndPlayer(input: SelectEdgeInput, inputter: UztaPlayer) {
		val edge = getUnownedLineSegment(input, inputter)

		val edgeCost = getCost(edge)
		val resourcesLeft = inputter.getResourcesLeftAfter(edgeCost)
		if (resourcesLeft.hasLackingItems()) {
			val neededResources = resourcesLeft.stream()
					.filter { pair -> pair.right < 0 }
					.map { this.presentNeededResources(it) }
					.collect(Collectors.joining(", "))
			throw UnacceptableInputException("You can''t afford that edge.  You still need $neededResources.")
		}

		setOwner(edge, inputter)
		inputter.giveResources(edgeCost.inverse)
	}

	private fun presentNeededResources(pair: Pair<NodeType, Int>): String {
		val numberNeeded = -pair.right
		return pair.left.getNumber(numberNeeded)
	}

	private fun getCost(edge: LineSegment): BaseIntegerBag<NodeType> {
		val bag = HashBag<NodeType>()
		edge.forEach { node -> bag.modifyNumberOf(node.type!!, node.values.size) }
		Arrays.stream(NodeType.values()).forEach { bag.addOne(it) }
		return bag
	}
}
